src/bugpoint.rs

```diff
@@ -0,0 +1,423 @@
+        reduce(isa, func, verbose);
+    }
+
+    Ok(())
+}
+
+enum MutationKind {
+    /// The mutation reduced the amount of instructions or ebbs.
+    Shrinked,
+    /// The mutation only changed an instruction. Performing another round of mutations may only
+    /// reduce the test case if another mutation shrank the test case.
+    Changed,
+}
+
+trait Mutator {
+    fn name(&self) -> &'static str;
+
+    fn mutation_count(&self, func: &Function) -> Option<usize>;
+
+    fn mutate(&mut self, func: Function) -> Option<(Function, String, MutationKind)>;
+
+    fn reduce(
+        &mut self,
+        isa: &TargetIsa,
+        mut func: Function,
+        progress_bar_prefix: String,
+        verbose: bool,
+        should_keep_reducing: &mut bool,
+    ) -> Function {
+        let progress =ProgressBar::with_draw_target(
+            self.mutation_count(&func).unwrap_or(0) as u64,
+            ProgressDrawTarget::stdout(),
+        );
+        progress.set_style(
+            ProgressStyle::default_bar()
+                .template("{bar:60} {prefix:40} {pos:>4}/{len:>4} {msg}"),
+        );
+
+        progress.set_prefix(&(progress_bar_prefix + &format!(" phase {}", self.name())));
+
+        for _ in 0..10000 {
+            progress.inc(1);
+
+            let (mutated_func, msg, mutation_kind) = match self.mutate(func.clone()) {
+                Some(res) => res,
+                None => {
+                    break;
+                }
+            };
+
+            progress.set_message(&msg);
+
+            match check_for_crash(isa, &mutated_func) {
+                CheckResult::Succeed => {
+                    // Shrinking didn't hit the problem anymore, discard changes.
+                    continue;
+                }
+                CheckResult::Verifier(_err) => {
+                    // Shrinking produced invalid clif, discard changes.
+                    continue;
+                }
+                CheckResult::Panic => {
+                    // Panic remained while shrinking, make changes definitive.
+                    func = mutated_func;
+                    match mutation_kind {
+                        MutationKind::Shrinked => {
+                            *should_keep_reducing = true;
+                            if verbose {
+                                progress.println(format!("{}: shrink", msg));
+                            }
+                        }
+                        MutationKind::Changed => {
+                            if verbose {
+                                progress.println(format!("{}: changed", msg));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        progress.set_message("done");
+        progress.finish();
+
+        func
+    }
+}
+
+/// Try to remove instructions.
+struct RemoveInst {
+    ebb: Ebb,
+    inst: Inst,
+}
+
+impl RemoveInst {
+    fn new(func: &Function) -> Box<dyn Mutator> {
```

Comment:
nit: This is an unexpected signature for `new` function in Rust. Maybe move the boxing outside of this function.